components/brave_federated_learning/brave_operational_profiling.cc

```diff
@@ -0,0 +1,220 @@
+
+}  // anonymous namespace
+
+BraveOperationalProfiling::BraveOperationalProfiling(
+    PrefService* pref_service,
+    scoped_refptr<network::SharedURLLoaderFactory> url_loader_factory)
+    : local_state_(pref_service), url_loader_factory_(url_loader_factory) {}
+
+BraveOperationalProfiling::~BraveOperationalProfiling() {}
+
+void BraveOperationalProfiling::RegisterLocalStatePrefs(
+    PrefRegistrySimple* registry) {
+  registry->RegisterIntegerPref(kLastCheckedSlotPrefName, -1);
+  registry->RegisterStringPref(kCollectionIdPrefName, {});
+  registry->RegisterTimePref(kCollectionIdExpirationPrefName, base::Time());
+}
+
+void BraveOperationalProfiling::Start() {
+  DCHECK(!simulate_local_training_step_timer_);
+  DCHECK(!collection_slot_periodic_timer_);
+
+  LoadPrefs();
+  InitPrefChangeRegistrar();
+  MaybeResetCollectionId();
+
+  simulate_local_training_step_timer_ =
+      std::make_unique<base::RetainingOneShotTimer>();
+  simulate_local_training_step_timer_->Start(
+      FROM_HERE,
+      base::TimeDelta::FromSeconds(
+          operational_profiling::features::
+              GetSimulateLocalTrainingStepDurationValue() *
+          60),
+      this, &BraveOperationalProfiling::OnSimulateLocalTrainingStepTimerFired);
+
+  collection_slot_periodic_timer_ = std::make_unique<base::RepeatingTimer>();
+  collection_slot_periodic_timer_->Start(
+      FROM_HERE,
+      base::TimeDelta::FromSeconds(
+          operational_profiling::features::GetCollectionSlotSizeValue() * 60 /
+          2),
+      this, &BraveOperationalProfiling::OnCollectionSlotStartTimerFired);
+}
+
+void BraveOperationalProfiling::Stop() {
+  simulate_local_training_step_timer_.reset();
+  collection_slot_periodic_timer_.reset();
+}
+
+void BraveOperationalProfiling::InitPrefChangeRegistrar() {
+  local_state_change_registrar_.Init(local_state_);
+  local_state_change_registrar_.Add(
+      brave::kP3AEnabled,
+      base::BindRepeating(&BraveOperationalProfiling::OnPreferenceChanged,
+                          base::Unretained(this)));
+}
+
+void BraveOperationalProfiling::LoadPrefs() {
+  last_checked_slot_ = local_state_->GetInteger(kLastCheckedSlotPrefName);
+  collection_id_ = local_state_->GetString(kCollectionIdPrefName);
+  collection_id_expiration_time_ =
+      local_state_->GetTime(kCollectionIdExpirationPrefName);
+}
+
+void BraveOperationalProfiling::SavePrefs() {
+  local_state_->SetInteger(kLastCheckedSlotPrefName, last_checked_slot_);
+  local_state_->SetString(kCollectionIdPrefName, collection_id_);
+  local_state_->SetTime(kCollectionIdExpirationPrefName,
+                        collection_id_expiration_time_);
+}
+
+void BraveOperationalProfiling::OnPreferenceChanged(const std::string& key) {
+  bool p3a_enabled = local_state_->GetBoolean(brave::kP3AEnabled);
+  bool is_operational_profiling_enabled =
+      operational_profiling::features::IsOperationalProfilingEnabled();
+  if (!p3a_enabled || !is_operational_profiling_enabled) {
+    Stop();
```

Comment:
I guess we also should restart it once again if the pref is on?

Reply:
👍